File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        ListNode[] ans = new ListNode[k];
        int len = 0;
        ListNode curr = head, prev = null;
        while(curr != null) {
            len++;
            curr = curr.next;
        } 

        int n = len/k, rem = len % k;
        curr = head;
        for(int i = 0; i<k; i++) {
           if(curr != null) {
                int sz = (rem > 0) ? 1 : 0;
                sz += n;
                ans[i] = curr;
                for(int j = 0; j<sz; j++) {
                    prev = curr;
                    curr = curr.next;
                }
                prev.next = null;
                rem--;
           }
           else {
               ans[i] = null;
           }
        }

        return ans;
    }
}

File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        int len = 0;
        ListNode* curr = head, *prev;
        while(curr) {
            len++;
            curr = curr -> next;
        } 

        int sz = len/k, rem = len % k;
        vector<ListNode*> ans(k);
        curr = head;
        
        for(int i = 0; i<k; i++) {
            if(curr) {
               ans[i] = curr;
               for(int j = 0; j<sz + (rem > 0); j++) {
                  prev = curr;
                  curr = curr -> next;
               }
               rem--;
               prev -> next = NULL;
            }
            else {
               ans[i] = NULL;
            }
        }

        return ans;
    }
};